### PR TITLE
FOCUS #185: Key-Value Format

### PR DESCRIPTION
**Pull request by kk09v**
Second iteration of the Key-Value Format after much discussion and lots of help from the people mentioned below

Co-authored-by: @flanakin
Co-authored-by: @rileyjenk
Co-authored-by: @cnharris10
Co-authored-by: @udam-f2

### CODE REVIEW

File: specification/attributes/key_value_format.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+## Description
+
+Rules and formatting requirements for columns appearing in billing data which convey data as key-value pairs.
+
+## Requirements
+
+* Columns subject to Key-Value Format MUST be a serialized JSON string, consistent with the ECMA-404 definition of an object.
+* Keys MUST be unique within an object.
+* Values MAY be a number, string, `true`, `false`, or `null`.
+  * Values MUST NOT be an object nor an array.
```
Comment (cnharris10) [inline specification/attributes/key_value_format.md:24]:
```suggestion
 * Values MUST NOT be an object nor an array.
```

File: specification/attributes/key_value_format.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Key-Value Format
+
+This is the specified format for providing data in the form of key-value pairs, which are also known as name-value pairs, attribute-value pairs, and field-value pairs in other contexts. 
+
+This format is often used in place of enumerating data which would be inherently sparse and/or without predetermined keys (attributes) into separate columns. To faciliate parsing, this format does not support nesting.
+
```
Comment (cnharris10) [inline specification/attributes/key_value_format.md:6]: From your initial paragraphs, tried to a bit less technical.  Feel free to judge as you wish ...

The Key-Value Format specifies the format for providing data as key-value pairs, also known as name-value pairs, attribute-value pairs, and field-value pairs in other contexts.

The Key-Value Format mirrors JSON's ECMA-404 with a handful or extra requirements and in most cases, more efficiently consolidates column-based data than the CSV format.

Comment (cnharris10) [reply specification/attributes/key_value_format.md:6]: Also removed this line "To faciliate parsing, this format does not support nesting." because it seemed duplicated in the requirements below

**Review by cnharris10 — CHANGES_REQUESTED** [review]

File: specification/attributes/key_value_format.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+## Description
+
+Rules and formatting requirements for columns appearing in billing data which convey data as key-value pairs.
+
+## Requirements
+
+* Columns subject to Key-Value Format MUST be a serialized JSON string, consistent with the ECMA-404 definition of an object.
+* Keys MUST be unique within an object.
+* Values MAY be a number, string, `true`, `false`, or `null`.
```
Comment (cnharris10) [inline specification/attributes/key_value_format.md:23]: Another option: `Values MUST be either a ...`

File: specification/attributes/key_value_format.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+Key-Value Format
+
+## Description
+
+Rules and formatting requirements for columns appearing in billing data which convey data as key-value pairs.
+
+## Requirements
+
+* Columns subject to Key-Value Format MUST be a serialized JSON string, consistent with the ECMA-404 definition of an object.
```
Comment (udam-f2) [inline specification/attributes/key_value_format.md:21]:
```suggestion
* Columns of type Key-Value Format MUST contain serialized JSON string, consistent with the ECMA-404 definition of an object.
```

File: specification/attributes/key_value_format.md (hunk 3 above)
Comment (udam-f2) [inline specification/attributes/key_value_format.md:23]: Is it clear that this is referring to the Key and Value of each KV pair defined in this column? Line 23 for example could get confused with the overall value of the cell? (I could be misreading that)

File: specification/attributes/key_value_format.md (hunk 2 above)
Comment (udam-f2) [reply specification/attributes/key_value_format.md:6]: @cnharris10 Not seeing where this is mentioned in requirements

File: specification/attributes/key_value_format.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Key-Value Format
+
+This is the specified format for providing data in the form of key-value pairs, which are also known as name-value pairs, attribute-value pairs, and field-value pairs in other contexts.
```
Comment (udam-f2) [inline specification/attributes/key_value_format.md:3]: For consistency, you might consider adding a sentence like we do for other format attributes where we explain why this would be helpful to consumers.

File: specification/attributes/key_value_format.md (hunk 2 above)
Comment (udam-f2) [inline specification/attributes/key_value_format.md:6]: To be consistent with the other attributes, we should add a normative line here that says MUST meet all requirements listed below (in the requirements section)

Comment (cnharris10) [reply specification/attributes/key_value_format.md:6]: Referring to this line, `* Values MUST NOT be an object nor an array.`.  If this is true, nesting is not possible.

**Review by cnharris10 — APPROVED** [review]

File: specification/attributes/key_value_format.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+Key-Value Format
+
+## Description
+
+Rules and formatting requirements for columns appearing in billing data which convey data as key-value pairs.
+
+## Requirements
+
+* Key-Value Format columns MUST contain a serialized JSON string, consistent with the ECMA-404 definition of an object.
```
Comment (joshk0) [inline specification/attributes/key_value_format.md:21]: Do we want to have an opinion on case sensitivity? As it stands this is permitting "keynull" and "KeYNuLl". Some CSP's have case sensitive tags, some CSP's don't.

**Review by joshk0 — COMMENTED** [review]
Nothing to stop shipping it, but some light edge cases I was able to think of

**Review by salpaysenturus — APPROVED** [review]

File: specification/attributes/key_value_format.md (hunk 6 above)
Comment (joshk0) [reply specification/attributes/key_value_format.md]: never mind. it is probably use case dependent (e.g. this is a discussion better left for `Tags`)

File: supporting_content/attributes/key_value_format.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Key-Value Format
+
+## Examples
+
+- Valid key-value format demonstrating all 5 valid values: `{"KeyString":"Value1", "KeyNumber":42, "KeyBool":true, "KeyBool2":false, "KeyEmpty":"", "KeyNull":null}`
```
[the remark this answers is not in the document]
Comment (joshk0) [reply supporting_content/attributes/key_value_format.md:5]: same thing, this is really about `Tags`

**Review by joshk0 — APPROVED** [review]

File: specification/attributes/key_value_format.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+# Key-Value Format
+
+Columns that provide Key-Value information are often used in place of separate columns for enumerating data which would be inherently sparse and/or without predetermined keys. This consolidates related information and provides more consistency in the schema. Key-value pairs are also referred to as name-value pairs, attribute-value pairs, or field-value pairs.
+
+All key-value related columns defined in the FOCUS specification MUST follow the key-value formatting requirements listed below. These columns SHALL NOT contain nested elements.
```
Comment (udam-f2) [inline specification/attributes/key_value_format.md:5]: Is the last sentence needed here? It's one of the requirements listed in the requirements already.

File: specification/attributes/key_value_format.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+## Attribute ID
+
+KeyValueFormat
+
+## Attribute Name
+
+Key-Value Format
```
Comment (udam-f2) [inline specification/attributes/key_value_format.md:13]: Should it be KeyValue Format? 

I ask because we also have Date/Time Format and it feels like it would be better to not have - or /. Otherwise we should pick one Applies here and for the top of the page header (see [here](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/specification/attributes/datetime_format.md))

Comment (ljadvey) [reply specification/attributes/key_value_format.md:13]: I like Keyvalue, DateTime....

File: specification/attributes/key_value_format.md (hunk 8 above)
Comment (ljadvey) [reply specification/attributes/key_value_format.md:5]: No reason to duplicate

File: specification/attributes/key_value_format.md (hunk 9 above)
Comment (flanakin) [reply specification/attributes/key_value_format.md:13]: Wikipedia uses [key-value](https://en.wikipedia.org/wiki/Key%E2%80%93value_database) and [date/time](https://en.wikipedia.org/wiki/ISO_8601#EDTF) just to keep us on our toes 🙃 I believe "DateTime" is slightly different in that it refers to a datatype in a programming language. I vaguely recall looking that up when we went thru that one.

**Review by flanakin — APPROVED** [review]

File: specification/attributes/key_value_format.md (hunk 9 above)
Comment (kk09v) [reply specification/attributes/key_value_format.md:13]: To me, the Attribute Name is written in plain English as you would when using it in an email and Attribute ID is written in the way you'd use it in a system or reference it in a slack message with technical teams.

In English, it's very atypical to have a compound word written in Pascal case.

For now, we can leave it as key-value since we have it written as date/time in the working draft. If we want to change this, we should do both (and any similar examples we find) in the same PR.

File: specification/attributes/key_value_format.md (hunk 8 above)
Comment (kk09v) [reply specification/attributes/key_value_format.md]: It was intended to make this more approachable, but the common opinion is that this redundant comment doesn't add value.

**Review by joshk0 — APPROVED** [review]

**Review by udam-f2 — APPROVED** [review]
LGTM.
